src/main/kotlin/eu/europa/ec/eudi/pidissuer/security/DPoPTokenReactiveAuthenticationManager.kt

```diff
@@ -0,0 +1,137 @@
+import org.springframework.security.core.Authentication
+import org.springframework.security.oauth2.core.OAuth2AuthenticatedPrincipal
+import org.springframework.security.oauth2.core.OAuth2AuthenticationException
+import org.springframework.security.oauth2.core.OAuth2TokenIntrospectionClaimNames
+import org.springframework.security.oauth2.server.resource.introspection.BadOpaqueTokenException
+import org.springframework.security.oauth2.server.resource.introspection.SpringReactiveOpaqueTokenIntrospector
+import reactor.core.publisher.Mono
+import reactor.kotlin.core.publisher.switchIfEmpty
+
+/**
+ * [ReactiveAuthenticationManager] implementing DPoP authentication.
+ * Implementation information can be found
+ * [here](https://connect2id.com/products/nimbus-oauth-openid-connect-sdk/examples/oauth/dpop#rs).
+ */
+class DPoPTokenReactiveAuthenticationManager(
+    private val introspector: SpringReactiveOpaqueTokenIntrospector,
+    private val verifier: DPoPProtectedResourceRequestVerifier,
+) : ReactiveAuthenticationManager {
+
+    /**
+     * Performs DPoP authentication.
+     */
+    override fun authenticate(authentication: Authentication): Mono<Authentication> =
+        Mono.just(authentication)
+            .filter { it is DPoPTokenAuthentication }
+            .map { it as DPoPTokenAuthentication }
+            .flatMap { dPoPAuthentication ->
+                introspect(dPoPAuthentication.accessToken)
+                    .flatMap { principal ->
+                        val issuer = principal.issuer()
+                        val thumbprint = principal.jwkThumbprint()
+                        Mono.zip(issuer, thumbprint)
+                            .flatMap {
+                                verify(dPoPAuthentication, it.t1, it.t2)
+                                    .then(Mono.just(dPoPAuthentication.authenticate(principal)))
+                            }
+                    }
+            }
+
+    /**
+     * Introspects the provided [token] and verifies whether it's active or not.
+     */
+    private fun introspect(token: DPoPAccessToken): Mono<OAuth2AuthenticatedPrincipal> =
+        introspector.introspect(token.value)
+            .onErrorMap { exception ->
+                val error =
+                    if (exception is BadOpaqueTokenException) {
+                        DPoPTokenError.invalidToken("Access token is not valid")
+                    } else {
+                        DPoPTokenError.serverError("Unable to introspect access token")
+                    }
+                OAuth2AuthenticationException(error, exception)
+            }
+
+    /**
+     * Given a [DPoPTokenAuthentication], the [DPoPIssuer], and the [JWKThumbprintConfirmation], performs DPoP verification
+     * using [verifier].
+     */
+    private fun verify(
+        authentication: DPoPTokenAuthentication,
+        issuer: DPoPIssuer,
+        thumbprint: JWKThumbprintConfirmation,
+    ): Mono<Unit> =
+        Mono.fromCallable {
```

Comment:
Can we remove the Mono from implementation?
Use `mono()` which lifts a suspended lamda into Mono context instead.